<?xml version="1.0"?>
<launch>
  <!-- Arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="map_file" default="$(find turtlebot3_navigation)/maps/map.yaml"/>
  <arg name="open_rviz" default="true"/>
  <arg name="move_forward_only" default="false"/>
  <arg name="waypoints_file" default="/home/masaya/maps/waypoints_willow.yaml"/>
  <arg name="dist_err" default="0.8"/>
  <arg name="initial_pose_x" default="-10.78" />
  <arg name="initial_pose_y" default="-16.78" />
  <arg name="initial_pose_a" default="0.0" />
  <arg name="use_waypoint_nav" default="false" />
  <arg name="rate" default="3.0"/>
  <arg name="loop" default="true"/>

  <!-- Turtlebot3 -->
  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)" />
  </include>

  <!-- Map server -->
  <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)"/>

  <!-- AMCL -->
  <include file="$(find turtlebot3_navigation)/launch/amcl.launch">
    <arg name="initial_pose_x" value="$(arg initial_pose_x)" />
    <arg name="initial_pose_y" value="$(arg initial_pose_y)" />
    <arg name="initial_pose_a" value="$(arg initial_pose_a)" />
  </include>

  <!-- move_base -->
  <include file="$(find turtlebot3_navigation)/launch/move_base.launch">
    <arg name="model" value="$(arg model)" />
    <arg name="move_forward_only" value="$(arg move_forward_only)"/>
    <arg name="cmd_vel_topic" value="/nav_vel"/>
  </include>

  <group unless="$(arg use_waypoint_nav)">
    <node name="waypoints_nav" pkg="fulanghua_waypoints_nav" type="waypoints_nav" output="screen">
      <param name="filename" value="$(arg waypoints_file)"/>
      <param name="dist_err" value="$(arg dist_err)"/>
    </node>
  </group>
  <group if="$(arg use_waypoint_nav)">
    <node name="waypoint_nav" pkg="waypoint_nav" type="waypoint_navigator_with_direction" output="screen">
      <param name="filename" value="$(arg waypoints_file)"/>
      <param name="dist_err" value="$(arg dist_err)"/>
      <param name="max_update_rate" value="$(arg rate)"/>
      <param name="loop_flag" value="$(arg loop)"/>    </node>
  </group>

  <!-- rviz -->
  <group if="$(arg open_rviz)"> 
    <node pkg="rviz" type="rviz" name="rviz" required="true"
          args="-d $(find turtlebot3_navigation)/rviz/turtlebot3_navigation.rviz"/>
  </group>
</launch>
